### PR TITLE
feat: support numeric and BIT C types in SQLBindParameter for string SQL types

### PR DESCRIPTION
**Pull request by sfc-gh-makowalski**
## Summary

- Expand `SnowflakeVarchar::ReadODBC` to properly handle integer, float, and BIT C types when bound to string SQL types (`SQL_VARCHAR`, `SQL_CHAR`, `SQL_LONGVARCHAR`, `SQL_WCHAR`, `SQL_WVARCHAR`, `SQL_WLONGVARCHAR`)
- Previously all non-WCHAR C types fell through to `read_char_str`, which interpreted raw struct bytes as a char string — producing garbage for numeric C types like `SQL_C_LONG` or `SQL_C_DOUBLE`
- Map `SQL_C_DEFAULT` to `read_char_str` to match the `WriteODBCType` side and avoid a regression from the prior catch-all
- Normalize `-0.0` to `"0"` for float-to-varchar binding to match old driver behavior
- The catch-all now returns `UnsupportedCDataType` error instead of silently misinterpreting unknown C types

### C types now supported for string SQL parameters

| C Type | Conversion |
|--------|-----------|
| `SQL_C_LONG` / `SQL_C_SLONG` | `i32 -> .to_string()` |
| `SQL_C_SHORT` / `SQL_C_SSHORT` | `i16 -> .to_string()` |
| `SQL_C_SBIGINT` | `i64 -> .to_string()` |
| `SQL_C_ULONG` | `u32 -> .to_string()` |
| `SQL_C_USHORT` | `u16 -> .to_string()` |
| `SQL_C_UBIGINT` | `u64 -> .to_string()` |
| `SQL_C_TINYINT` / `SQL_C_STINYINT` | `i8 -> .to_string()` |
| `SQL_C_UTINYINT` | `u8 -> .to_string()` |
| `SQL_C_DOUBLE` | `f64 -> .to_string()` (negative zero normalized to "0") |
| `SQL_C_FLOAT` | `f32 -> .to_string()` (negative zero normalized to "0") |
| `SQL_C_BIT` | `u8 -> "0" or "1"` |

### Follow-up (PR 2)

Structured C types (`SQL_C_TYPE_TIMESTAMP`, `SQL_C_TYPE_DATE`, `SQL_C_TYPE_TIME`, `SQL_C_NUMERIC`, `SQL_C_BINARY`) are handled in the stacked PR #749.

## Test plan

- [x] 36 e2e tests:
  - Each numeric C type bound to SQL_VARCHAR (SLONG, ULONG, SSHORT, USHORT, SBIGINT, UBIGINT, STINYINT, UTINYINT, DOUBLE, FLOAT)
  - Negative value test (negative SLONG)
  - BIT true, BIT false, BIT value > 1
  - Deprecated C type aliases (SQL_C_LONG, SQL_C_SHORT, SQL_C_TINYINT)
  - SQL_C_SLONG bound to all 5 other string SQL types (SQL_CHAR, SQL_LONGVARCHAR, SQL_WCHAR, SQL_WVARCHAR, SQL_WLONGVARCHAR)
  - Boundary values: INT_MIN, INT_MAX, zero, LLONG_MIN, LLONG_MAX, ULLONG_MAX
  - Float edge cases: -0.0, NaN (double + float), +infinity (double + float), -infinity (double + float)
  - SQL_C_DEFAULT bound to SQL_VARCHAR
- [x] All existing unit tests continue to pass

SNOW-3031761

### CODE REVIEW

**Comment by sfc-gh-makowalski** [comment]
* **#768** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/768?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#749** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/749?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#748** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/748?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/748?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: odbc/src/conversion/varchar.rs [hunk 1]
```diff
@@ -380,8 +380,31 @@ impl ReadODBC for SnowflakeVarchar {
...
         let s = match binding.value_type {
+            CDataType::Char => read_char_str(binding)?,
             CDataType::WChar => read_wchar_str(binding)?,
-            _ => read_char_str(binding)?,
+            CDataType::Long | CDataType::SLong => read_unaligned::<i32>(binding).to_string(),
+            CDataType::Short | CDataType::SShort => read_unaligned::<i16>(binding).to_string(),
+            CDataType::SBigInt => read_unaligned::<i64>(binding).to_string(),
+            CDataType::ULong => read_unaligned::<u32>(binding).to_string(),
+            CDataType::UShort => read_unaligned::<u16>(binding).to_string(),
+            CDataType::UBigInt => read_unaligned::<u64>(binding).to_string(),
```
[the remark this answers is not in the document]
Comment (sfc-gh-makowalski) [reply odbc/src/conversion/varchar.rs:390]: The features might diverge, no point in creating a shared helper

File: odbc/src/conversion/varchar.rs [hunk 2]
```diff
@@ -380,8 +380,31 @@ impl ReadODBC for SnowflakeVarchar {
...
+                    "1".to_string()
+                } else {
+                    "0".to_string()
+                }
+            }
+            _ => {
+                return UnsupportedCDataTypeSnafu {
+                    c_type: binding.value_type,
+                }
+                .fail();
```
[the remark this answers is not in the document]
Comment (sfc-gh-makowalski) [reply odbc/src/conversion/varchar.rs:420]: Fixed

File: odbc_tests/tests/e2e/query/sql_bind_parameter.cpp [hunk 3]
```diff
@@ -728,3 +728,392 @@ TEST_CASE("should report APD count zero when no parameters are bound.", "[query]
...
+// =============================================================================
+// C type → VARCHAR conversions
+// =============================================================================
+
+TEST_CASE("should bind SQL_C_LONG to SQL_VARCHAR.", "[query][bind_parameter][c_to_varchar]") {
+  // Given Snowflake client is logged in
+  Connection conn;
+  auto stmt = conn.createStatement();
+  SQLINTEGER param = 42;
+  SQLLEN indicator = 0;
```
Comment (sfc-gh-pbulawa) [inline odbc_tests/tests/e2e/query/sql_bind_parameter.cpp]: For all the tests, let's set the indicator to the fixed size of the type for consistency, e.g., `sizeof(param)`

File: odbc_tests/tests/e2e/query/sql_bind_parameter.cpp
Comment (sfc-gh-pbulawa) [inline odbc_tests/tests/e2e/query/sql_bind_parameter.cpp:1]: **Please add following tests:**

1. **Boundary values for integer types.** No test covers `INT_MIN`, `INT_MAX`, `LLONG_MIN`, `LLONG_MAX`, `ULLONG_MAX`, or zero for any numeric type.
2. **Negative test for unsupported C types.** The new catch-all returns `UnsupportedCDataType`, but no test verifies this error path works (e.g., binding `SQL_C_TYPE_TIMESTAMP` to `SQL_VARCHAR)`.
3. **Float edge cases.** The `f64::to_string()` / `f32::to_string()` path is tested with only exact-representable values (`3.14`, `1.5`). Missing cases:  
\- Values requiring scientific notation (e.g., `1e300`, `1e-300`) — does Snowflake accept `"1e300"` as a VARCHAR parameter value?  
\- `NAN` , `INFINITY``  
`\- `-0.0`
4. **BIT with values > 1.**
5. **`SQL_C_DEFAULT`** **for** **`SQL_VARCHAR`**

Comment (sfc-gh-pbulawa) [inline odbc_tests/tests/e2e/query/sql_bind_parameter.cpp:1]: We should restructure the binding tests to follow a similar conversion for the sql-to-c tests as it will benefit in readibility, however that's not a part of this PR

**Review by sfc-gh-pbulawa — CHANGES_REQUESTED** [review]

File: odbc_tests/tests/e2e/query/sql_bind_parameter.cpp (hunk 3 above)
Comment (sfc-gh-makowalski) [reply odbc_tests/tests/e2e/query/sql_bind_parameter.cpp]: 👍

File: odbc_tests/tests/e2e/query/sql_bind_parameter.cpp
Comment (sfc-gh-makowalski) [reply odbc_tests/tests/e2e/query/sql_bind_parameter.cpp:1]: 1. 👍
2. 👍
3. Working on it, but seems not so trivial
4. 👍
5. 👍

Comment (sfc-gh-makowalski) [reply odbc_tests/tests/e2e/query/sql_bind_parameter.cpp:1]: Ad. 2. Turns out that all C types should be supported, if not now then in the following PR, so I'll skip this test if we would remove it anyway in the future. source: https://learn.microsoft.com/en-us/sql/odbc/reference/appendixes/converting-data-from-c-to-sql-data-types?view=sql-server-ver17  
Ad. 3. I see you already noticed the differences and documented them in float.cpp file about nan/NaN capitalization. I'll add tests with similar workarounds

File: odbc/src/conversion/varchar.rs [hunk 4]
```diff
@@ -380,8 +380,45 @@ impl ReadODBC for SnowflakeVarchar {
...
+                let v = read_unaligned::<f64>(binding);
+                if v == 0.0 {
+                    0.0_f64.to_string()
+                } else {
+                    v.to_string()
+                }
+            }
+            CDataType::Float => {
+                let v = read_unaligned::<f32>(binding);
+                if v == 0.0 {
```
Comment (sfc-gh-pbulawa) [inline odbc/src/conversion/varchar.rs:403]: Is this if branch necessary?

File: odbc/src/conversion/varchar.rs [hunk 5]
```diff
@@ -380,8 +380,45 @@ impl ReadODBC for SnowflakeVarchar {
...
+            CDataType::Short | CDataType::SShort => read_unaligned::<i16>(binding).to_string(),
+            CDataType::SBigInt => read_unaligned::<i64>(binding).to_string(),
+            CDataType::ULong => read_unaligned::<u32>(binding).to_string(),
+            CDataType::UShort => read_unaligned::<u16>(binding).to_string(),
+            CDataType::UBigInt => read_unaligned::<u64>(binding).to_string(),
+            CDataType::TinyInt | CDataType::STinyInt => read_unaligned::<i8>(binding).to_string(),
+            CDataType::UTinyInt => read_unaligned::<u8>(binding).to_string(),
+            CDataType::Double => {
+                let v = read_unaligned::<f64>(binding);
+                if v == 0.0 {
```
Comment (sfc-gh-pbulawa) [inline odbc/src/conversion/varchar.rs:395]: Is this if branch necessary?

File: odbc_tests/tests/e2e/query/sql_bind_parameter.cpp
Comment (sfc-gh-makowalski) [reply odbc_tests/tests/e2e/query/sql_bind_parameter.cpp:1]: I'll try to do it in the following PR - 768

**Review by sfc-gh-pbulawa — APPROVED** [review]

File: odbc/src/conversion/varchar.rs (hunk 5 above)
Comment (sfc-gh-makowalski) [reply odbc/src/conversion/varchar.rs:395]: Yes — Rust's `(-0.0_f64).to_string()` produces `"-0"`, while the old driver (via Simba SDK) normalizes it to `"0"`. Without this, the server echoes back "-0" as VARCHAR and the test fails. IEEE 754 guarantees `-0.0 == 0.0` is true but `NaN == 0.0` is false, so this only affects negative zero.

File: odbc/src/conversion/varchar.rs (hunk 4 above)
Comment (sfc-gh-makowalski) [reply odbc/src/conversion/varchar.rs:403]: Same as above — normalizes `-0.0f` to `"0"` to match old driver behavior.
